### PR TITLE
IcingaDB: include consistent IDs, start time and end time in icinga:history:stream:{flapping,acknowledgement}

### LINKED ISSUE #7689: IcingaDB: Flapping & Acknowledgement - Track start/set timestamp
This is needed to allow Icinga DB to only insert one history row into MySQL

### PR DESCRIPTION
**Pull request by Al2Klimov**
fixes #7689

### CODE REVIEW

**Comment by Al2Klimov** [comment]
## Blocked by

* [ ] #7695

File: lib/icingadb/icingadb-objects.cpp [hunk 1]
```diff
@@ -1650,6 +1647,27 @@ void IcingaDB::SendFlappingChanged(const Checkable::Ptr& checkable, const Value&
...
+
+		xAdd.emplace_back("event_type");
+		xAdd.emplace_back("flapping_end");
+		xAdd.emplace_back("end_time");
+		xAdd.emplace_back(Convert::ToString(TimestampToMilliseconds(changeTime)));
+	}
+
+	xAdd.emplace_back("start_time");
+	xAdd.emplace_back(Convert::ToString(startTime));
+	xAdd.emplace_back("id");
```
Comment (N-o-X) [inline lib/icingadb/icingadb-objects.cpp]: This should be `flapping_id` instead of `id`

File: lib/icingadb/icingadb-objects.cpp [hunk 2]
```diff
@@ -1713,10 +1729,17 @@ void IcingaDB::SendAcknowledgementSet(const Checkable::Ptr& checkable, const Str
 		xAdd.emplace_back(Convert::ToString(TimestampToMilliseconds(expiry)));
 	}
 
+	long long setTime = TimestampToMilliseconds(changeTime);
+
+	xAdd.emplace_back("set_time");
+	xAdd.emplace_back(Convert::ToString(setTime));
+	xAdd.emplace_back("id");
```
Comment (N-o-X) [inline lib/icingadb/icingadb-objects.cpp]: This should be `acknowledgement_id` or `ack_id` instead of `id`

File: lib/icinga/clusterevents.cpp [hunk 3]
```diff
@@ -494,6 +494,7 @@ void ClusterEvents::AcknowledgementSetHandler(const Checkable::Ptr& checkable,
 	params->Set("notify", notify);
 	params->Set("persistent", persistent);
 	params->Set("expiry", expiry);
+	params->Set("change_time", changeTime);
```
Comment (N-o-X) [inline lib/icinga/clusterevents.cpp]: `change_time` should be `set_time` or `clear_time` depending on the event

File: lib/icinga/clusterevents.cpp [hunk 4]
```diff
@@ -558,6 +559,7 @@ void ClusterEvents::AcknowledgementClearedHandler(const Checkable::Ptr& checkabl
 	if (service)
 		params->Set("service", service->GetShortName());
 	params->Set("author", removedBy);
```
Comment (N-o-X) [inline lib/icinga/clusterevents.cpp]: This should not be `author` but something like `removed_by`

File: lib/icinga/clusterevents.cpp [hunk 5]
```diff
@@ -537,12 +538,12 @@ Value ClusterEvents::AcknowledgementSetAPIHandler(const MessageOrigin::Ptr& orig
 
 	checkable->AcknowledgeProblem(params->Get("author"), params->Get("comment"),
 		static_cast<AcknowledgementType>(static_cast<int>(params->Get("acktype"))),
-		params->Get("notify"), params->Get("persistent"), params->Get("expiry"), origin);
+		params->Get("notify"), params->Get("persistent"), params->Get("change_time"), params->Get("expiry"), origin);
```
Comment (dnsmichi) [inline lib/icinga/clusterevents.cpp]: This crashes if the instance receives a cluster message from <2.12 instances.

It needs a logical `if (params->Contains("change_time") ...`.

File: lib/icinga/apievents.cpp [hunk 6]
```diff
@@ -182,7 +182,7 @@ void ApiEvents::FlappingChangedHandler(const Checkable::Ptr& checkable, const Me
 
 void ApiEvents::AcknowledgementSetHandler(const Checkable::Ptr& checkable,
 	const String& author, const String& comment, AcknowledgementType type,
-	bool notify, bool persistent, double expiry, const MessageOrigin::Ptr& origin)
+	bool notify, bool persistent, double, double expiry, const MessageOrigin::Ptr& origin)
```
Comment (dnsmichi) [inline lib/icinga/apievents.cpp]: Where's the parameter? Shouldn't we declare and use that also in our REST API?

File: lib/icinga/checkable-flapping.cpp [hunk 7]
```diff
@@ -64,10 +64,16 @@ void Checkable::UpdateFlappingStatus(bool stateChange)
...
 	SetFlappingIndex(oldestIndex);
 	SetFlappingCurrent(flappingValue);
-	SetFlapping(flapping, true);
 
-	if (flapping != GetFlapping())
-		SetFlappingLastChange(Utility::GetTime());
+	if (flapping != GetFlapping()) {
+		SetFlapping(flapping, true);
+
+		double ee = GetLastCheckResult()->GetExecutionEnd();
```
Comment (dnsmichi) [inline lib/icinga/checkable-flapping.cpp]: Why the change from `now` to `execution_end`?

File: lib/icinga/checkable-flapping.cpp [hunk 8]
```diff
@@ -64,10 +64,16 @@ void Checkable::UpdateFlappingStatus(bool stateChange)
 	SetFlappingBuffer(stateChangeBuf.GetValue());
 	SetFlappingIndex(oldestIndex);
 	SetFlappingCurrent(flappingValue);
-	SetFlapping(flapping, true);
 
-	if (flapping != GetFlapping())
-		SetFlappingLastChange(Utility::GetTime());
+	if (flapping != GetFlapping()) {
+		SetFlapping(flapping, true);
```
Comment (dnsmichi) [inline lib/icinga/checkable-flapping.cpp]: Logic change, why?

**Comment by dnsmichi** [comment]
Also, please explain the purpose of `change_time`. This gets set whenever an acknowledgement is added and cleared. Isn't that a meta information for the event handlers knowing the details themselves? 

Likewise, the `change_time` would likely apply for any runtime state. Is that something more general to be discussed with @lippserd ?

File: lib/icinga/clusterevents.cpp (hunk 5 above)
Comment (Al2Klimov) [reply lib/icinga/clusterevents.cpp]: * Doesn't `Dictionary#Get()` return `Empty`?
* Isn't `double(Empty)` `0.0`?

File: lib/icinga/checkable-flapping.cpp (hunk 7 above)
Comment (Al2Klimov) [reply lib/icinga/checkable-flapping.cpp]: Each node calcs flapping – and "now" – by itself, but we need a consistent ID based on this across nodes.

File: lib/icinga/checkable-flapping.cpp (hunk 8 above)
Comment (Al2Klimov) [reply lib/icinga/checkable-flapping.cpp]: We listen for `SetFlapping()` and want to be only notified when it actually changes.

File: lib/icinga/apievents.cpp (hunk 6 above)
Comment (Al2Klimov) [reply lib/icinga/apievents.cpp]: IMAO it's an internal only info. In our ReST API we have `Checkable#acknowledgement_last_change` now.

**Comment by Al2Klimov** [comment]
`change_time` is about the same as now vs. `execution_end`.
